### DISCUSSION TITLE
Could MIR (eventually) be used to compile a language with tail call elimination?

### DISCUSSION THREAD

**Discussion by jgonggrijp** [vnmakarov/mir#409] [Q&A] [answered]
From the blog posts, the README and a few other discussions, I understand that MIR currently assumes the C calling convention. Given that tail call elimination (TCE) is not doable with that convention, that seems to suggest that TCE is, at least for now, out of reach for any compiler targeting MIR (at least without trampolining or complicated heuristic code transformations). Although the thing going on with thunks might create opportunities here? I don't understand enough about the inner workings of MIR, or compilers in general for that matter, to draw my own conclusions on that front.

Functional programming languages often like to employ tail call elimination as a loop primitive. Are such languages within the scope of MIR, or do you plan to support them in the future? What would be your advice to someone who wants to implement such a language and benefit from the high speed, low latency execution profile that MIR offers?

In any case, thank you for this very interesting project!

**Comment by jgonggrijp** [comment]
@vnmakarov and @derekbsnider please allow me to draw your attention to this topic.

**Comment by derekbsnider** [comment]
@jgonggrijp It is possible. I believe `gcc` does this with `-O2` optimization. This might also be something that you'd only want if optimizations were not disabled (`-O0`).

**Reply by jgonggrijp** [reply]
It does that, but that's only tail call *optimization*. Which works only in a subset of cases and is not universally applied as a result. If you try to make a loop by having a function tail-call itself, you can still easily cause a stack overflow. Also, even if it did work universally as an optimization, it would still be a problem without optimization. When debugging, you still don't want your loops to cause stack overflows.

(FWIW, I learned this the hard way by trying it with a toy programming language that I compiled to C++. The only way to make tail calls work reliably was to use trampolining.)

**Reply by derekbsnider** [reply]
I just mean there should probably be a reliable way to control whether or not this occurs... if this was something you wanted to be a standard/default way your language worked, then it would set this option, but I don't know for sure if you would want it to always happen, as I believe there are some cases where it wouldn't be safe to happen.

**Reply by jgonggrijp** [reply]
I don't know how this works in GCC exactly, but the way LLVM lets you control this is by offering different [calling conventions](https://llvm.org/docs/LangRef.html#calling-conventions). Hence my question.

**Reply by cyanogilvie** [reply]
I don't know if it's really applicable to c as more than an optimization, but some languages rely on it to work at all (as you mentioned with the functional approach to loops).  Tcl supports this with an explicit `tailcall` command but needs a trampoline to implement this on the c side.  Maybe I'm being naive but it doesn't seem like it would be too hard to do at the MIR level though (essentially swapping out the call frame from one `func` invocation with another, without `ret` from the first).  It would need implementation in the hardware lowering code to implement for the calling convention, but I don't see a structural reason that would prevent it.

**Reply by jgonggrijp** [reply]
Thanks for chiming in. Suppose I were to add this feature to MIR (currently not able to, but might in the future). Which parts of the code would I have to edit?

**Reply by cyanogilvie** [reply, marked as answer]
I've attached what Claude thinks will be required, which matches broadly what I would expect:
- new insn MIR_TAILCALL (parsing, API, serializing, validate that the return signature matches the current function's one).
- teach the optimization pipeline what tailcall means (essentially RET+CALL).
- implement in the interpreter (shouldn't be too difficult).
- get involved in each target calling convention's machine code lowering to implement as: function epilogue + directly chaining to the tailcall without transferring control back to the saved return address but leaving it as the return address for the tailcalled function.

An issue Claude flagged that I hadn't spotted is a restriction on the tailcalled function to not need more stack space for spilled arguments that don't fit in the calling convention's assigned argument registers (since our caller will adjust the stack pointer for what it pushed there when control eventually returns to it).

What's your use case?  The usual functional language pattern of implementing loops as tail recursive calls to the current function would get these two guarantees for free - it's the same function so the return signature and arg stack footprint match naturally.  If that's all you need then I don't even think any MIR changes are required - just assign the argument regs and MIR_JMP to the function top (before the locals init if your language has that).

If it needs more general TCE but has the invariant that the return and argument sig always matches on the mir-func-level (say, return is a single pointer to a `value` object and arguments are fixed as state pointer and list of `value` objects (as pointer-to-array + size)), then all "functions" in your language have compatible mir function signatures for TCE (this would lock you out of unboxing the `value` objects across call boundaries though).  I'm plotting to JIT Tcl with this, which happens to have this property: all commands are essentially `int objcmd(void* cdata, Tcl_Interp* interp, size_t objc, Tcl_Obj* objv[])`, so it's realistically possible even across functions with minimal MIR changes I think (release stack local allocas, replace `cdata`, `interp`, `objc` and `objv` arguments wherever they live in the calling convention, and jump to the top of the new function).  Or invent your own calling convention for internal calls and use the MIR_JCALL/MIR_JRET machinery.

If your use case lacks the invariant that mir-func signatures naturally match, then I think you could hack around the arg stack space issue by synthesising a dummy stack spilled arg (or args) to reserve enough space for any realistic arg signature (consumes more stack space for each function call frame, but since this is presumably in support of stackless function nesting through TCE, it really doesn't matter - only a single function's callframe will be on the stack at any given time).

Claude's survey (line number references are from my worktree which has patches and WIP not upstreamed yet, so they're likely to be a little off compared to current `master`):

---

My prompt (in case some misunderstanding on my part coloured Claude's assessment):

    What would be involved in adding support for tail call elimination to MIR, something
    like a MIR_TAILCALL version of MIR_CALL that replaces the current call frame with the
    called one, without returning, leaving its return address as the return address for the
    tail called function (whose return type would have to match)?

# Adding `MIR_TAILCALL` to MIR

A design/scoping note on what it would take to add proper tail-call elimination
to MIR: a `MIR_TAILCALL` variant of `MIR_CALL` that replaces the current call
frame with the callee's, without returning, leaving the caller's return address
as the return address for the tail-called function (whose return type must match
the caller's).

## The good news: MIR already has half of this

MIR already ships `MIR_JCALL`/`MIR_JRET` — a jump-based call that *doesn't* set
up a return address (the backend emits a `jmp`/`br` instead of `call`, and
`jret_p` functions get a frame with no return-pc slot). See:

- `MIR.md:503` — the documented semantics.
- `mir-gen-x86_64.c:1844` — the pattern-table rows (and the equivalents in every
  other backend: `mir-gen-aarch64.c:1756`, `mir-gen-ppc64.c:1442`,
  `mir-gen-riscv64.c:1756`, `mir-gen-s390x.c:1441`).
- `mir-gen-x86_64.c:34` — the frame layout comment ("absent for jcall/jret func").

But `JCALL`/`JRET` is **not** tail-call elimination. Per the docs it is a
deliberately non-standard ABI for **void, no-argument** functions, built for
switching a direct-threading interpreter into JITted code, passing values
through hard-reg-tied globals (`MIR.md:511-513`). The whole function is `jret_p`
(a per-function property set at `MIR_func_finish`, `mir.c:1622`), not a
per-call-site decision, and it carries no arguments or results.

So the JCALL machinery is the *precedent* and shows exactly which code paths a
tail call touches — but a real `MIR_TAILCALL` (general args, matching return
type, per-call-site) is a meaningfully bigger job because of argument
marshalling and frame teardown ordering.

## What a `MIR_TAILCALL` would have to guarantee

1. The callee's prototype return type(s) must match the caller's `func->nres` /
   result types exactly (so the callee's `RET` writes the values the caller's
   caller expects, in the caller's return registers).
2. Before the jump, the caller's frame must be fully torn down: callee-saved
   registers restored, `alloca` areas and stack-arg areas released, SP/FP reset
   to the *caller's caller* state — **except** the return address must be left
   in place (on stack or in the link register) so the callee returns past us.
3. The outgoing arguments must be placed into the ABI arg registers/stack slots,
   and the stack portion of those args has to land in the frame slot that the
   return address sits above — which is the part that overlaps with, and can
   clobber, the very frame you're tearing down.

## Concrete pieces to touch

### IR layer (`mir.h` / `mir.c`) — the cheap, mechanical part

- Add `MIR_TAILCALL` to the insn enum (`mir.h:136` region, the
  `REP3(INSN_EL, CALL, INLINE, JCALL)` line) and the insn descriptor table
  (`mir.c:330` area).
- A constructor (mirror `MIR_new_jcall` near `mir.c:2391`).
- Add it to `MIR_call_code_p` (`mir.h:466`).
- Validation in the func-finish pass (`mir.c:1620-1637`): a tailcall's prototype
  results must equal the function's own results; and decide the RET/JRET-mixing
  analogue.
- Text/binary reader-writer and `MIR_insn_nops` / op-mode handling.

### Generator core (`mir-gen.c`)

Every site that special-cases `MIR_RET`/`MIR_JRET` because a tailcall is *both* a
call and a block/function terminator. There are ~12 such sites (CFG construction
`9704`/`9727`, liveness `2060`/`6079`, dead-code `9293`, the "is this a BB end"
predicate `3830`, `5379`, `10091`, etc.). Each needs to treat `MIR_TAILCALL` as
a terminator that also has call clobber semantics. This is the part most likely
to bite — a tailcall is a control-flow leaf (no fallthrough) but also clobbers
caller-saved regs and uses args, a combination no existing insn has.

### Backends (`mir-gen-x86_64.c` + the other four) — the real work

- `machinize_call` (`mir-gen-x86_64.c:172`) already lowers args into arg regs /
  outgoing stack slots; you'd reuse it, but the outgoing stack-arg region for a
  tailcall must be positioned at the *caller's* incoming-arg area, not freshly
  pushed below SP. If the tailcall passes more stack args than the current frame
  received, you have to either grow into reserved space or fall back.
- The prologue/epilogue logic (`target_make_prologue_epilogue`, around
  `1155`–`1325`) currently keys frame teardown off `func->jret_p`. For a
  per-call-site tailcall you need an *epilogue-then-jump* sequence emitted at the
  tailcall: restore callee-saved regs (the loop at `1294`), `add sp, frame_size`,
  restore FP, but keep the return pc, then `jmp` to the target (reuse the
  `MIR_JCALL` pattern at `1844`). The ordering against argument materialization
  is the crux — args computed into temps before teardown, then moved into final
  arg slots after SP is restored.
- A pattern-table entry per backend (trivial — copy the `MIR_JCALL` rows).
- ABI restrictions to detect and reject (or fall back to a normal call): varargs
  callees, struct/`BLK` args passed on stack, `long double`, anything where the
  outgoing stack-arg footprint exceeds the incoming one. Simplest first cut:
  only allow tailcalls whose stack-arg size ≤ the current function's, matching
  result types, non-vararg.

### Interpreter (`mir-interp.c`)

If tailcalls must work interpreted (they should, for parity): add a
`CASE(MIR_TAILCALL)` (`1541` region) and an `IC_IMM_TAILCALL` immediate variant
(`60`-`68`, `437`, `1016`-`1022`). The interpreter doesn't have a real machine
frame, so semantically a tailcall is just "call then propagate the result as our
own return" — easy, but you must make it actually reuse the frame (loop instead
of recursing) or you lose the whole point (no C-stack growth).

### c2mir + optionally `[[musttail]]`

Not required for the IR feature, but if you want C-level access you'd emit
`MIR_TAILCALL` for `return f(args);` in tail position. That's a separate,
optional layer.

## The genuinely hard parts (in priority order)

1. **Argument/frame overlap on the stack.** Restoring SP and writing outgoing
   stack args touch the same memory. This is why production compilers restrict
   tail calls (e.g. LLVM `tailcc`/`musttail` requires matching signatures and
   bounded stack args). Start by only supporting the non-overlapping case.
2. **The "terminator that is also a call" duality in the generator** — auditing
   all ~12 RET/JRET branch sites in `mir-gen.c` so liveness and CFG stay correct.
3. **Five backends.** The pattern entry is trivial; the epilogue-before-jump
   sequencing is per-arch and where bugs hide (link-register architectures —
   aarch64/ppc64/riscv64/s390x — must avoid reloading the LR before the branch).

## Suggested scoping

- **Phase 1:** IR insn + validation + interpreter (loop-reuse) + text/binary IO.
  Gives a testable, correct-but-unoptimized feature.
- **Phase 2:** x86_64 generator only, restricted to the safe subset (matching
  results, non-vararg, outgoing stack args ≤ incoming). Reuse `machinize_call`
  + the `JCALL` jump pattern + a tailcall-specific epilogue.
- **Phase 3:** Port to the other four backends, then relax the stack-arg
  restriction.

**Reply by jgonggrijp** [reply]
Thanks for the elaborate answer!

I do need more than self-recursion. For example, a loop of sorts might also result from two different functions recursively calling each other. However, it is no problem to give each function the same signature, in fact I was already considering to do something like that.
